### ISSUE TITLE
Cache-based compile

### ISSUE THREAD

**Issue by cdriesler** [WeWorkSandbox/svgar#21]
Svgar currently redraws the entire composition when `compile()` is called, no matter how small the change. This is mostly fine, but is a performance issue when `compile()` needs to be rapidly called on moderate to large drawings (like, say, when trying to pan the camera).

Opt to refactor `compile()` to simply read and combine a cache of output scopes in the `SvgarCube` object. These caches should be flaggable at any moment so that they are re-computed the next time `compile()` is called. 

Scopes for first version:

- root
  - top level `svg` tag
  - `viewBox` as camera
  - `width` and `height` as div extents
- style
  - `#cubename > * ` with global styles
  - `#slabname > * ` with local styles
  - `#slabname > .stylename ` for specific path state styles
- defs
  - `g` tag with `id=<slab name>`
  - `path` tags with `class="<style in state>"` and `id=<uuid>`
- global
  - top level `g` tag for cube and global styles with `id=<cube name>`
- local
  - `<use x="<atX>" y="<-atY>" xlink:href="#slabname" />`
  - repeat for each instance of item placed

Compile then becomes:

`[root, style, defs, global, local, "</g>", "</svg>"].join("/n")`

This opens up a lot, but seems to also take a bit away. For example, I can't fake `:hover` for styles because [:hover does not exist for def elements](https://stackoverflow.com/questions/13673441/svg-use-element-and-hover-style). Still seems like a net positive for performance. Will re-evaluate :hover later.

**Comment by cdriesler** [comment]
New sketch schema:

```
<svg viewBox width height>

<defs>
    <g id=(layer)>
        <style>
        #layer > * { } <!--- local style --->
        #layer > .state { } <!--- stateful style --->
        </style>
        <path class=(state) d />
    </g>
</defs>

<g id=(cube)>
    <style>
    #cube > * { } <!--- global style --->
    </style>
    <use x y xlink:href=(slab) />
</g>

</svg>
```

This reduces the number of scopes to three:

- root `<svg>`
- slabs `<defs>`
- cube `<g>`

Each slab will cache/flag itself. The cube object will just coordinate their placement and usage (i.e. the x & y in use, or any masking).

**Comment by cdriesler** [comment]
Use of defs causes major performance issues (see 3f4fcd7 ). Revert to old schema + minor new changes:

```
<svg viewBox w h>

<style>
    <!-- global styles -->
    #cube > * { }

    <!-- local styles -->
    #slab > * { }

    <!-- stateful styles -->
    #slab > .state { }
</style>

<g id=(cube)>
    <g id=(slab)>
        <path vector-effect="non-scaling-stroke" class=(state) d />
    </g>
</g>

</svg>
```

Scopes are:

- root `<svg>`
- style `<style>`
- geometry `<g>`
  - note: each slab will still cache itself if possible

**Comment by cdriesler** [comment]
Implemented in `v0.3.25`
